### PR TITLE
#4978 Implement prv_backup_to_designated_mailbox client configuration rule

### LINKED ISSUE #4978: backup option that emails encrypted backup to a designated mailbox
with this option client configuration:

```
"prv_backup_to_designated_mailbox": "(armored public key here)"
```

User will be presented with a screen that we already have as in "used pgp before?" with options to import or create a key.

When importing, they have an option to import from a file or by pasting as usual.

When generating, after the user sets their pass phrase (the regular backup step), without asking the user anything regarding which backup to use, the key will be:

1) decrypted, so it doesn't use any pass phrase + armored
2) email will be composed, to: `<primary uid of the prv_backup_to_designated_mailbox pubkey>` , subject: `FlowCrypt OpenPGP Private Key backup for user ${userEmail} with id ${primaryFingerprint}`, the text should have: `Please keep this email and attachment in the mailbox for safekeeping. It will be needed if the user ever needs to set up FlowCrypt again or forgets their pass phrase. Without this, the user won't be able to read their emails. See https://flowcrypt.com/docs/technical/enterprise/configuration/backup-to-designated-mailbox.html#recovery` and then the attachment will be the private key
3) This whole email will be encrypted with the public key in `prv_backup_to_designated_mailbox`
4) the email will be sent to the primary uid email contained in `prv_backup_to_designated_mailbox` public key

The priority is as follows, which should be documented when adding this new option to https://flowcrypt.com/docs/technical/enterprise/configuration/client-configuration.html#client-configuration-rules : 
 - if both `prv_backup_to_designated_mailbox` and `key_manager_url` is set, `prv_backup_to_designated_mailbox` will be ignored
 - if both `prv_backup_to_designated_mailbox` and `NO_PRV_BACKUP` is set, `prv_backup_to_designated_mailbox` takes precedence 
 - if both `PRV_AUTOIMPORT_OR_AUTOGEN` and `prv_backup_to_designated_mailbox` are set, client will show error

### PR DESCRIPTION
**Pull request by martgil**
This PR implements the `prv_backup_to_designated_mailbox` client configuration org rule.

close #4978

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: extension/js/common/ui/backup-ui/backup-ui-manual-module.ts [hunk 1]
```diff
@@ -63,6 +65,28 @@ export class BackupUiManualActionModule extends BackupUiModule<BackupUi> {
     }
   };
 
+  public doBackupOnDesignatedMailbox = async (
+    acctEmail: string,
```
Comment (sosnovsky) [inline extension/js/common/ui/backup-ui/backup-ui-manual-module.ts:69]: probably `acctEmail` param isn't needed and we can use `this.ui.acctEmail` instead?

File: extension/chrome/settings/setup/setup-create-key.ts [hunk 2]
```diff
@@ -35,7 +40,56 @@ export class SetupCreateKeyModule {
...
+              onBackedUpFinished: async () => {
+                $('pre.status_details').remove();
+                $('#backup-template-container').remove();
+                await this.view.finalizeSetup();
+                await this.view.setupRender.renderSetupDone();
+              },
+            });
+            const parsedPrivateKey = await KeyUtil.parse(privateKey[0].private);
+            await OpenPGPKey.decryptKey(parsedPrivateKey, opts.passphrase);
+            const armoredPrivateKey = KeyUtil.armor(parsedPrivateKey);
```
Comment (sosnovsky) [inline extension/chrome/settings/setup/setup-create-key.ts:65]: It looks like you armor encrypted key, as result of `OpenPGPKey.decryptKey` isn't used anywhere

File: test/source/tests/setup.ts [hunk 3]
```diff
@@ -2299,6 +2303,180 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+        /* eslint-disable @typescript-eslint/naming-convention */
+        const settingsPage = await BrowserRecipe.openSettingsLoginApprove(t, browser, acctEmail);
+        await SetupPageRecipe.createKey(settingsPage, '', 'email', {
+          skipForPassphrase: true,
+          backupPrvKeyToMailbox: true,
+          key: {
+            passphrase: 'super difficult to guess passphrase',
+          },
+        });
+        const expectedEmailSubject = `FlowCrypt OpenPGP Private Key backup for user`;
```
Comment (sosnovsky) [inline test/source/tests/setup.ts:2333]: we can probably check full email subject, as we have user email and key fingerprint

File: test/source/tests/setup.ts [hunk 4]
```diff
@@ -2299,6 +2303,180 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+        const decryptedAttachmentContent = await MsgUtil.decryptMessage({
+          kisWithPp: kisWithPassphrase,
+          encryptedData: encryptedAttachmentContent,
+          verificationPubs: [pubkey],
+        });
+        expect(fromEmailHeaderLine.includes(acctEmail)).to.equal(true);
+        expect(toEmailHeaderLine.includes(adminEmail)).to.equal(true);
+        expect(subjectLine.includes(acctEmail)).to.equal(true);
+        expect(decrypted.success).to.equal(true);
+        expect(decryptedAttachmentContent.success).to.equal(true);
```
Comment (sosnovsky) [inline test/source/tests/setup.ts:2362]: Let's test full process - you should also check if decrypted attachment content equals to decrypted user's private key, and then try to setup extension with this key and new passphrase.
It's very important functionality, so should be tested well

File: test/source/tests/setup.ts [hunk 5]
```diff
@@ -2299,6 +2303,180 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+        const settingsPage = await BrowserRecipe.openSettingsLoginApprove(t, browser, acctEmail);
+        await SetupPageRecipe.createKey(settingsPage, '', 'disabled', {
+          skipForPassphrase: true,
+          backupPrvKeyToMailbox: true,
+          key: {
+            passphrase: 'super difficult to guess passphrase',
+          },
+        });
+        const expectedEmailSubject = `FlowCrypt OpenPGP Private Key backup for user`;
+        const sentMsg = (await GoogleData.withInitializedData(acctEmail)).searchMessagesBySubject(expectedEmailSubject)[0];
```
Comment (sosnovsky) [inline test/source/tests/setup.ts:2393]: this test is almost the same as `Client configuration prv_backup_to_designated_mailbox test`, let's just check if message with `expectedEmailSubject` was sent, it should be enough here

File: extension/chrome/settings/setup/setup-create-key.ts [hunk 6]
```diff
@@ -35,7 +40,56 @@ export class SetupCreateKeyModule {
...
       const keyIdentity = await this.createSaveKeyPair(opts, keyAlgo);
-      if (this.view.clientConfiguration.canBackupKeys()) {
+      if (this.view.clientConfiguration.getPublicKeyForPrivateKeyBackupToDesignatedMailbox()) {
+        const adminPubkey = this.view.clientConfiguration.getPublicKeyForPrivateKeyBackupToDesignatedMailbox();
+        if (adminPubkey) {
+          const msgEncryptionKey = await KeyUtil.parse(adminPubkey);
+          const destinationEmail = msgEncryptionKey.emails[0];
+          try {
+            const privateKey = await KeyStore.get(this.view.acctEmail);
+            const primaryKeyId = await privateKey[0].id;
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/settings/setup/setup-create-key.ts]: corrected. thank you.

File: extension/js/common/ui/backup-ui/backup-ui-manual-module.ts (hunk 1 above)
Comment (martgil) [reply extension/js/common/ui/backup-ui/backup-ui-manual-module.ts]: oh - yes, can remove the passed parameter and use `this.ui.acctEmail`.

File: extension/chrome/settings/setup/setup-create-key.ts [hunk 7]
```diff
@@ -35,7 +40,56 @@ export class SetupCreateKeyModule {
...
+                await this.view.finalizeSetup();
+                await this.view.setupRender.renderSetupDone();
+              },
+            });
+            const parsedPrivateKey = await KeyUtil.parse(privateKey[0].private);
+            await OpenPGPKey.decryptKey(parsedPrivateKey, opts.passphrase);
+            const armoredPrivateKey = KeyUtil.armor(parsedPrivateKey);
+            const encryptedPrivateKey = await MsgUtil.encryptMessage({
+              pubkeys: [msgEncryptionKey],
+              data: await Buf.fromUtfStr(armoredPrivateKey),
```
[the remark this answers is not in the document]
Comment (martgil) [reply extension/chrome/settings/setup/setup-create-key.ts]: the mistake is corrected. thank you.

File: extension/chrome/settings/setup/setup-create-key.ts (hunk 2 above)
Comment (sosnovsky) [reply extension/chrome/settings/setup/setup-create-key.ts:64]: Hmm, maybe you're right, implementing test for restoration process (https://github.com/FlowCrypt/flowcrypt-browser/pull/5313#discussion_r1288361666) will be useful for checking it

File: test/source/tests/setup.ts [hunk 8]
```diff
@@ -2299,6 +2303,180 @@ AN8G3r5Htj8olot+jm9mIa5XLXWzMNUZgg==
...
+          fes: {
+            clientConfiguration: {
+              flags: ['NO_ATTESTER_SUBMIT'],
+              prv_backup_to_designated_mailbox: pubkey,
+              enforce_keygen_algo: 'ecc25519',
+              enforce_keygen_expire_months: 6,
+            },
+          },
+        });
+        /* eslint-disable @typescript-eslint/naming-convention */
```
[the remark this answers is not in the document]
Comment (martgil) [reply test/source/tests/setup.ts]: updated. thanks.

File: extension/chrome/settings/setup/setup-create-key.ts (hunk 2 above)
Comment (martgil) [reply extension/chrome/settings/setup/setup-create-key.ts:64]: I understand. will do it.

File: test/source/tests/setup.ts (hunk 5 above)
Comment (martgil) [reply test/source/tests/setup.ts:2392]: updated. thanks.

File: test/source/tests/setup.ts (hunk 3 above)
Comment (martgil) [reply test/source/tests/setup.ts]: yes - I have updated the expectedEmailSubject to include the user and key fingerprint.

File: test/source/tests/setup.ts (hunk 4 above)
Comment (martgil) [reply test/source/tests/setup.ts:2368]: I have added a test for setting up user's account using the backup private key. thank you.

**Comment by martgil** [comment]
@sosnovsky ready for review. thank you.

**Review by sosnovsky — APPROVED** [review]
Well done 👍

**Pull request merged**